crates/services/gas_price_service/src/fuel_gas_price_updater/fuel_da_source_adapter.rs

```diff
@@ -1,16 +1,145 @@
 use crate::fuel_gas_price_updater::{
-    DaCommitDetails,
-    DaCommitSource,
+    DaGasPrice,
+    Error::CouldNotFetchDARecord,
+    GetDaGasPriceFromSink,
     Result as GasPriceUpdaterResult,
+    SetDaGasPriceToSink,
 };
+use anyhow::anyhow;
+use std::sync::Arc;
+use tokio::sync::Mutex;
 
-#[derive(Default, Clone)]
-pub struct FuelDaSource;
+pub mod block_committer_source;
+pub mod dummy_source;
+pub mod service;
 
-impl DaCommitSource for FuelDaSource {
-    fn get_da_commit_details(
-        &mut self,
-    ) -> GasPriceUpdaterResult<Option<DaCommitDetails>> {
-        todo!() // TODO(#2139): pending research on how to get the data from the block committer
+pub use block_committer_source::BlockCommitterDaGasPriceSource;
+pub use dummy_source::DummyDaGasPriceSource;
+pub use service::*;
+
+pub const POLLING_INTERVAL_MS: u64 = 10_000;
+
+pub type DaGasPriceProvider = Arc<Mutex<Option<DaGasPrice>>>;
+
+impl GetDaGasPriceFromSink for DaGasPriceProvider {
+    fn get(&mut self) -> GasPriceUpdaterResult<Option<DaGasPrice>> {
+        let mut gas_price_guard = self.try_lock().map_err(|err| {
+            CouldNotFetchDARecord(anyhow!(
+                "Failed to lock shared gas price state: {:?}",
+                err
+            ))
+        })?;
+
+        let da_gas_price = gas_price_guard.clone();
+
+        // now mark it as consumed because we don't want to serve the same data
+        // multiple times
+        *gas_price_guard = None;
+
+        Ok(da_gas_price)
+    }
+}
+
+impl SetDaGasPriceToSink for DaGasPriceProvider {
+    fn set(&mut self, value: DaGasPrice) -> GasPriceUpdaterResult<()> {
+        let mut gas_price_guard = self.try_lock().map_err(|err| {
+            CouldNotFetchDARecord(anyhow!(
+                "Failed to lock shared gas price state: {:?}",
+                err
+            ))
+        })?;
+
+        *gas_price_guard = Some(value);
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::fuel_gas_price_updater::{
+        fuel_da_source_adapter::service::Result as DaGasPriceSourceResult,
+        DaGasPriceProviderService,
+        DaGasPriceSource,
+        DummyDaGasPriceSource,
+    };
+    use fuel_core_services::{
+        RunnableService,
+        Service,
+        ServiceRunner,
+    };
+    use std::time::Duration;
+    use tokio::time::sleep;
+
+    #[derive(Default)]
+    struct ErroringSource;
+
+    #[async_trait::async_trait]
+    impl DaGasPriceSource for ErroringSource {
+        async fn get(&mut self) -> DaGasPriceSourceResult<DaGasPrice> {
+            Err(anyhow!("boo!"))
+        }
+    }
+
+    type TestValidService =
+        DaGasPriceProviderService<DummyDaGasPriceSource, DaGasPriceProvider>;
+    type TestErroringService =
+        DaGasPriceProviderService<ErroringSource, DaGasPriceProvider>;
+
+    #[tokio::test]
+    async fn test_service_sets_cache_when_request_succeeds() {
```

Comment:
nit: I think we should be a little clearer on what the SUT of these tests is. I think it's safe to say that it's actually `run`, not `start`? The service runner is just calling `run` for us.
```suggestion
    async fn run__when_da_block_cost_source_gives_value_shared_value_is_updated() {
```

Here is a WIP doc on testing standards that talks about this a little:
https://www.notion.so/fuellabs/Engineering-Style-Guide-83ba6fd81f12443796c814ecfc746b39?pvs=4#f3232125ea964e959a66b9f41267d96e

Reply:
addressed in https://github.com/FuelLabs/fuel-core/pull/2163/commits/74fc9bf4e08cb0d4a1a5def23f31d2c4fbd31c9f#diff-d0929c84f197be22ea7a8f1bc32981a49731a6ec12250819e9c86ebba1373ffaR71